upkie/cpp/actuation/BulletInterface.cpp

```diff
@@ -353,6 +361,41 @@ void BulletInterface::read_joint_sensors() {
     result.torque = sensor_state.m_jointMotorTorque;
   }
 }
+void BulletInterface::save_nominal_masses() {
+  const int nb_links = bullet_.getNumJoints(robot_);
+  b3DynamicsInfo info;
+  for (int link_id = 0; link_id < nb_links; ++link_id) {
+    bullet_.getDynamicsInfo(robot_, link_id, &info);
+    nominal_masses[link_id] = info.m_mass;
+    nominal_inertia[link_id][0] = info.m_localInertialDiagonal[0];
+    nominal_inertia[link_id][1] = info.m_localInertialDiagonal[1];
+    nominal_inertia[link_id][2] = info.m_localInertialDiagonal[2];
+  }
+}
+void BulletInterface::randomize_masses() {
+  for (const auto& link_id : nominal_masses) {
+    std::default_random_engine generator;
+    std::uniform_real_distribution<double> distribution(-inertia_randomization_,
+                                                        inertia_randomization_);
+    double epsilon = distribution(generator);
+    RobotSimulatorChangeDynamicsArgs change_dyn_args;
+    change_dyn_args.m_mass = nominal_masses[link_id.first] * (1 + epsilon);
+    std::uniform_real_distribution<double> distribution01(0, 1);
+    double epsilon0 = distribution01(generator);
+    double epsilon1 = distribution01(generator);
+    double epsilon2 = distribution01(generator);
+    epsilon0 = epsilon * epsilon0 / (epsilon0 + epsilon1 + epsilon2);
+    epsilon1 = epsilon * epsilon1 / (epsilon0 + epsilon1 + epsilon2);
```

Comment:
Sorry, my bad for pushing to this more complex solution. Actually this one doesn't work, as it violates one property of one of the references I linked in the previous review. Consider for instance the case where:

```math
\epsilon_0 > \epsilon_1 \quad \wedge \quad \epsilon_0 > \epsilon_2
```

Reply:
I think your initial guess was simpler (thus better, unless we identify a limitation) and correct:

- Mass density $\rho = m / V$, total mass divided by body volume
- We update the total mass to $m' = (1 + \epsilon) m$
- Thus $\rho' = m' / V$ is the new mass density: since it is $\rho' = (1 + \epsilon) \rho$, and inertia is linear in mass, we can simply multiply the original inertia by $1 + \epsilon$ without recomputing

You could even add an inline comment to inform future travelers who'd jump to conclusions too fast like I did :sweat_smile: